dipy/io/statefull_tractogram.py

```diff
@@ -0,0 +1,879 @@
+from copy import deepcopy
+import enum
+from itertools import product
+import logging
+import numbers
+import os
+import time
+import six
+
+import nibabel as nib
+from nibabel.affines import apply_affine
+from nibabel.streamlines.array_sequence import ArraySequence
+from nibabel.streamlines import detect_format
+from nibabel.streamlines.tractogram import (Tractogram,
+                                            PerArraySequenceDict,
+                                            PerArrayDict)
+import numpy as np
+
+from dipy.io.vtk import save_vtk_streamlines, load_vtk_streamlines
+from dipy.io.dpy import Dpy
+
+
+class Space(enum.Enum):
+    """ Enum to simplify future change to convention """
+    VOX = 'vox'
+    VOXMM = 'voxmm'
+    RASMM = 'rasmm'
+
+
+class StateFullTractogram(object):
+    """ Object designed to be identical no matter the file format
+    (trk, tck, fib, dpy). Facilitate transformation between space and
+    data manipulation for each streamline / point.
+    """
+
+    def __init__(self, streamlines, reference, space,
+                 shifted_origin=False,
+                 data_per_point=None, data_per_streamline=None):
+        """ Create a strict, state-aware, robust tractogram
+
+        Parameters
+        ----------
+        streamlines : list or ArraySequence
+            Streamlines of the tractogram
+        reference : Nifti or Trk filename, Nifti1Image or TrkFile,
+            Nifti1Header or trk.header (dict)
+            Reference that provides the spatial attribute.
+            Typically a nifti-related object from the native diffusion used for
+            streamlines generation
+        space : string
+            Current space in which the streamlines are (vox, voxmm or rasmm)
+            Typically after tracking the space is VOX, after nibabel loading
+            the space is RASMM
+        shifted_origin : bool
+            Information on the position of the origin,
+            False is Trackvis standard, default (corner of the voxel)
+            True is NIFTI standard (center of the voxel)
+        data_per_point : dict
+            Dictionary in which each key has X items, each items has Y_i items
+            X being the number of streamlines
+            Y_i being the number of points on streamlines #i
+        data_per_streamline : dict
+            Dictionary in which each key has X items
+            X being the number of streamlines
+
+        Notes
+        -----
+        Very important to respect the convention, verify that streamlines
+        match the reference and are effectively in the right space.
+
+        Any change to the number of streamlines, data_per_points or
+        data_per_streamlines requires particular verification.
```

Comment:
Well I meant to say that changes to these attribute (using the function of the class) will fail if invalid (failing to respect the length mentioned in the description).
What is the optimal way to 'warn' the user that these attributes need to remain consistent with each other ?
(I ask, because I am a bit too much close to the code to know what is clear)